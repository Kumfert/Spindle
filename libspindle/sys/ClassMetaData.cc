//
// ClassMetaData.h
//
// $Id: ClassMetaData.cc,v 1.2 2000/02/18 01:32:00 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//

#include "spindle/ClassMetaData.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

SpindleBaseClass*
ClassMetaData::createObject() {
  if (pfnCreateObject == 0) { 
    return 0;
  } else { 
    return (*pfnCreateObject)();
  }
}

bool 
ClassMetaData::isDerivedFrom( const ClassMetaData* p ) const {
  for ( const ClassMetaData* parent = this; parent != 0; parent = parent->parentMetaData ) {
    if ( parent == p ) { 
      return true;
    }
  }
  return false;
}
