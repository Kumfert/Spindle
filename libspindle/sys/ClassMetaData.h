//
// ClassMetaData.h
//
// $Id: ClassMetaData.h,v 1.2 2000/02/18 01:32:00 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
//=====================================================================
//

#ifndef SPINDLE_CLASS_META_DATA_H_
#define SPINDLE_CLASS_META_DATA_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class SpindleBaseClass;

/** 
 *  @memo Meta Data on descendants of SpindleBaseClass
 *  @type class
 * 
 *  @author Gary Kumfert
 *  @version #$Id: ClassMetaData.h,v 1.2 2000/02/18 01:32:00 kumfert Exp $#
 *  @see SpindleBaseClass
 */

class ClassMetaData {
public:

  /// A zero terminated string of the class's name
  char* className;

  /// The size of the class as returned by #sizeof()#
  int  classSize;

  /// A pointer to the meta data of the class's parent
  ClassMetaData *parentMetaData;

  /** a pointer to the default constructor of the class
   * Can be NULL if the class is abstract
   */
  SpindleBaseClass* (* pfnCreateObject )(); 

  /// The count of number of instances created.
  int nCreated;

  /// The count of number of instances destroyed.
  int nDestroyed;
 
  /**
   * @memo creates a new instance of the class the metadata represents.
   * @return a new instance if it can create one, else NULL.
   */
  SpindleBaseClass* createObject();

  /**
   * @memo Used before #dynamic_cast<>()# was available
   * @return true if this meta data represents a descendent of the
   *         class represented by p.
   * @param p non-null pointer to a class's metadata.
   */
  bool isDerivedFrom( const ClassMetaData* p) const;
};

SPINDLE_END_NAMESPACE
 
#endif
