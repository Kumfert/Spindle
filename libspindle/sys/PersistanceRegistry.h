//
// PersistanceRegistry.h -- 
//
// $Id: PersistanceRegistry.h,v 1.2 2000/02/18 01:32:01 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//

#ifndef SPINDLE_PERSISTANCE_REGISTRY_H_
#define SPINDLE_PERSISTANCE_REGISTRY_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifdef REQUIRE_OLD_CXX_HEADER_SUFFIX
#include <stdio.h>
#include "map.h"
#else
#include <cstdio>
#include <map>
using namespace std;
#endif

#ifndef SPINDLE_FUNC_H_
#include "spindle/spindle_func.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class SpindleSystem;
class ClassMetaData;

class PersistanceRegistry { 
  friend SpindleSystem;
public:
  typedef map< const char*, const ClassMetaData*, str_less > RegistryType;

private:
  RegistryType registry;
  PersistanceRegistry( const PersistanceRegistry& src ); // no implementation
  void operator=( const PersistanceRegistry& src );      // no implementation
  PersistanceRegistry();
  FILE * dumpOnDestroy;

public:
  static PersistanceRegistry& get();

  ~PersistanceRegistry();
  void registerClass( const ClassMetaData* metaData );
  const ClassMetaData* findClass( const char* className );
  void dump( FILE * stream ) const;  
  void setDumpOnDestroy( FILE* fp) { dumpOnDestroy = fp; }
  
};

SPINDLE_END_NAMESPACE

#endif 
