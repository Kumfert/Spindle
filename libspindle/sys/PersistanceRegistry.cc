//
// PersistanceRegistry
//
// $Id: PersistanceRegistry.cc,v 1.2 2000/02/18 01:32:01 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//

#include "spindle/PersistanceRegistry.h"

#ifndef SPINDLE_CLASS_META_DATA_H_
#include "spindle/ClassMetaData.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

PersistanceRegistry::PersistanceRegistry() {
  dumpOnDestroy = 0;
}

PersistanceRegistry::~PersistanceRegistry() {
  if ( dumpOnDestroy != 0 ) {
    fprintf( dumpOnDestroy,"Destroying Registry... ");
    dump( dumpOnDestroy );
  }
}


PersistanceRegistry& 
PersistanceRegistry::get() {
  static PersistanceRegistry pr;
  return pr;
}

void
PersistanceRegistry::registerClass( const ClassMetaData* metaData ) {
  if ( registry.find( metaData->className ) == registry.end() ) { // if its not yet registered
    registry[ metaData->className ] = metaData;
  }
}
 
const ClassMetaData*
PersistanceRegistry::findClass( const char* className ) {
  RegistryType::iterator it = registry.find( className );
  if ( it != registry.end() ) { // we've found it!
    return (*it).second;
  }
  return 0;
}

void
PersistanceRegistry::dump( FILE * stream ) const {
 
#ifdef SPINDLE__NO_INSTANCE_COUNT
  fprintf(stream ,"WARNING: Instance counting disabled for this compilation platform\n");
#endif

 fprintf(stream,"PersistanceRegistery::dump():  %d items\n", registry.size());
  for( RegistryType::const_iterator it = registry.begin(), stop = registry.end();
       it != stop; ++it ) {
    const ClassMetaData* self = (*it).second;
    if ( self != 0 ) {
      const char* className = ( (self->className == 0) ? "unnamed" : self->className );
      const char* parentName = "";
      const ClassMetaData* parent = self->parentMetaData;
      if ( parent != 0 ) {
	parentName = ( (parent->className == 0) ? parentName : parent->className);
      }
      fprintf(stream, "  %25s (%d) %d/%d---> %s\n", className, self->classSize, self->nCreated, 
	      self->nDestroyed, parentName);
    } // end if (self != 0)
  } // end for-all in registry
}



