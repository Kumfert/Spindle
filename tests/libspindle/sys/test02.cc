//
// src/sys/drivers/tests/test02.cc
//
// $Id: test02.cc,v 1.2 2000/02/18 01:32:11 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
// tests the PersistanceRegistry
//

#include "spindle/PersistanceRegistry.h"

#ifndef SPINDLE_CLASS_META_DATA_H_
#include "spindle/ClassMetaData.h"
#endif 


#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

static ClassMetaData SpindleBaseClass_MetaData =  { "SpindleBaseClass", 12, 0, 0, 0, 0 };
static ClassMetaData SpindlePersistant_MetaData = { "SpindlePersistant", 16, &SpindleBaseClass_MetaData, 0, 0, 0 };
static ClassMetaData SpindleFile_MetaData = { "SpindleFile", 20, &SpindleBaseClass_MetaData, 0, 0, 0 };
static ClassMetaData SpindleStdioFile_MetaData = { "SpindleStdioFile", 20, &SpindleFile_MetaData, 0, 0, 0 };
static ClassMetaData MatrixFile_MetaData = { "MatrixFile", 50, &SpindleStdioFile_MetaData, 0, 0, 0 };
static ClassMetaData ChacoFile_MetaData = { "ChacoFile", 50, &MatrixFile_MetaData, 0, 0, 0 };
static ClassMetaData HarwellBoeingFile_MetaData = { "HarwellBoeingFile", 80, &MatrixFile_MetaData, 0, 0, 0 };
static ClassMetaData IJValFile_MetaData = { "IJValFile", 50, &MatrixFile_MetaData, 0, 0, 0 };
static ClassMetaData MatrixMarketFile_MetaData = { "MatrixMarketFile", 50, &MatrixFile_MetaData, 0, 0, 0 };

#define REGISTRY PersistanceRegistry::get()

int main() {
  REGISTRY.registerClass( &SpindleBaseClass_MetaData );
  REGISTRY.registerClass( &SpindlePersistant_MetaData );
  REGISTRY.registerClass( &SpindleFile_MetaData );
  REGISTRY.registerClass( &SpindleStdioFile_MetaData );
  REGISTRY.registerClass( &MatrixFile_MetaData );
  REGISTRY.registerClass( &HarwellBoeingFile_MetaData );
  REGISTRY.registerClass( &MatrixMarketFile_MetaData);
  REGISTRY.registerClass( &ChacoFile_MetaData );
  REGISTRY.registerClass( &IJValFile_MetaData );
  
#ifdef SPINDLE__NO_INSTANCE_COUNT
  fprintf(stdout,"WARNING: Instance counting disabled for this compilation platform\n");
#endif
  REGISTRY.dump(stdout);

  fprintf(stdout,"So, what is the parent of \"ChacoFile\"?...\n");
  const ClassMetaData* metaData = REGISTRY.findClass("ChacoFile");
  fprintf(stdout,"\t why, its \"%s\".\n", metaData->parentMetaData->className );
  fprintf(stdout,"And how big is a HarwellBoeingFile object?\n");
  metaData = REGISTRY.findClass("HarwellBoeingFile");
  fprintf(stdout,"\t why, its %d bytes.\n", metaData->classSize );

  
}
