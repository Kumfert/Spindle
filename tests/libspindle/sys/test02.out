PersistanceRegistery::dump():  9 items
                  ChacoFile (50) 0/0---> MatrixFile
          HarwellBoeingFile (80) 0/0---> MatrixFile
                  IJValFile (50) 0/0---> MatrixFile
                 MatrixFile (50) 0/0---> SpindleStdioFile
           MatrixMarketFile (50) 0/0---> MatrixFile
             SpindleBaseClass (12) 0/0---> 
                  SpindleFile (20) 0/0---> SpindleBaseClass
            SpindlePersistant (16) 0/0---> SpindleBaseClass
             SpindleStdioFile (20) 0/0---> SpindleFile
So, what is the parent of "ChacoFile"?...
	 why, its "MatrixFile".
And how big is a HarwellBoeingFile object?
	 why, its 80 bytes.
